test/e2e/network_segmentation_custom_l2udn.go

```diff
@@ -0,0 +1,138 @@
+			RequiredUDNNamespaceLabel: "",
+		})
+		f.Namespace = namespace
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	DescribeTable("should respect network configuration",
+		func(netConfig *networkAttachmentConfigParams) {
+			By("creating the L2 network")
+
+			netConfig.namespace = f.Namespace.Name
+			udnManifest := generateUserDefinedNetworkManifest(netConfig)
+			cleanup, err := createManifest(netConfig.namespace, udnManifest)
+			Expect(err).NotTo(HaveOccurred())
+			DeferCleanup(cleanup)
+			Eventually(userDefinedNetworkReadyFunc(f.DynamicClient, netConfig.namespace, netConfig.name), 5*time.Second, time.Second).Should(Succeed())
+
+			By("creating a pod on the custom L2 network")
+			podConfig := *podConfig("gateway-test-pod")
+			podConfig.namespace = f.Namespace.Name
+			pod := runUDNPod(cs, f.Namespace.Name, podConfig, nil)
+
+			By("getting the created UDN object to validate against")
+			udn, err := udnClient.K8sV1().UserDefinedNetworks(netConfig.namespace).Get(context.Background(), netConfig.name, metav1.GetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+			Expect(udn.Spec.Layer2).NotTo(BeNil())
+
+			By(fmt.Sprintf("validating gateway configuration for pod %s", pod.Name))
+			podAnno, err := unmarshalPodAnnotation(pod.Annotations, pod.Namespace+"/"+netConfig.name)
+			Expect(err).NotTo(HaveOccurred())
+
+			expectedGatewayIPs := make(map[string]bool)
+			if len(udn.Spec.Layer2.DefaultGatewayIPs) > 0 {
+				for _, gwIP := range udn.Spec.Layer2.DefaultGatewayIPs {
+					expectedGatewayIPs[string(gwIP)] = true
+				}
+			} else {
+				// without custom gateways default values are expected
+				for _, subnet := range udn.Spec.Layer2.Subnets {
+					expectedGatewayIPs[util.GetNodeGatewayIfAddr(testing.MustParseIPNet(string(subnet))).IP.String()] = true
+				}
+			}
+
+			expectedGWs := len(udn.Spec.Layer2.Subnets)
+			for _, route := range podAnno.Routes {
+				if route.NextHop != nil {
+					gwIP := route.NextHop.String()
+					if expectedGatewayIPs[gwIP] {
+						expectedGWs--
+						By(fmt.Sprintf("Found custom gateway IP %s in pod routes", gwIP))
+						if expectedGWs == 0 {
+							break
+						}
+					}
+				}
+			}
+			Expect(expectedGWs).To(BeZero(), "Custom gateway IPs should be found in pod routes")
+
+			// Check that pod IP is not in reserved CIDRs (if any are configured)
+			if len(udn.Spec.Layer2.ReservedSubnets) > 0 {
+				By(fmt.Sprintf("validating pod %s is not in reserved subnet", pod.Name))
+				for _, ip := range podAnno.IPs {
+					podIP := ip.IP.String()
+					fmt.Println(podIP)
+					// Check against each reserved subnet
+					for _, reservedSubnet := range udn.Spec.Layer2.ReservedSubnets {
+						_, reservedNet, err := net.ParseCIDR(string(reservedSubnet))
+						Expect(err).NotTo(HaveOccurred())
+						Expect(reservedNet.Contains(net.ParseIP(podIP))).To(BeFalse(),
+							fmt.Sprintf("Pod IP %s should not be in reserved subnet %s", podIP, reservedSubnet))
+					}
+				}
+			}
```

Comment:
ditto

Reply:
This is only checking that the POD IP doesn't belong to the reserved subnet, what's the alternative you propose?